[WIP] OptimAE

Halbaroth opened the pull request:
This PR is mostly a rebase of the branch `optims`: https://github.com/OCamlPro/alt-ergo/compare/next...optims.
All the commits between e392d5b071f0414b78ea552693dc4b8bec9ed46a and c4b336c0cba5383275808fe215dde00cca1feb2f have been merged by the PR #530 

- I prevent `ocp-indent` to reindent the mly files. It seems there is a unfixed bug and the parser was a mess after running ocp-indent on it. See https://github.com/OCamlPro/ocp-indent/issues/218

Notice that I also ignored these commits: 
- c4b336c0cba5383275808fe215dde00cca1feb2f
- 3f01bf60eb0178964c83d9ee2f55bf4622c82782 (old code)
- 3c52d274bfb7247feb2f13a8d780160791dda75a (old code)
- fca77c9f6fc883794be0ffadef6c82442a1f282e (old code)
- 57d4af580b2edef18ab1554b071d3a11b7fd37a9 (old code)
- 78811d80b05e3f33d34de8c012dca5b1a77ca51d (already applied)
- 90bb216dca1146c9f3b3f9f140959a83f6c1e4e3 (already applied)
- 6405133eac1780a011809c73a5f51a4687a2c2bb (already applied)
- e8e6c2d3656888739aa04417956b24787cbccaa0 (old code)
- cb2e74aeb291bef28c4a5e9b55c0cf049d98ee6c (poetry)
- f7c655d2350625610d85384aaa1a51bfa4d9faea (old code)
- e87fb9d84270fe4edd50c4bc01a49daa9ee3b984 (already applied)
- 202c7684c4396cf1713b0ca722e1707fbc76a426 (already applied)
- b534dc957daa877ae2dabfcc68ceda80982b7eac (already applied)
- 6b344b6f4956c41998cf73c017209a1d5e279270 (already applied)
- ce50c902266fcfd3d2cdcdfdfe6c0f713e578319 (already applied)
- 71e7feeaa171c15e1d55d00baf74012b934c423a (already applied)
- 0a1e3307fb81d3d401be7ef980445b44d4506131 (already applied)
- 2523faef0ca35b52c73fd7c6f3070083da0bffb3 (already removed)
- 86876467ca9f71099a66ba534c39445c467a932c (already applied)
- e541e5ccf510bd444c3a82e9d1a6222f0825cbf1 (already applied)
- 400f3fd4ed300de4359df80bf99b97ea0bbf2763 (poetry)
- a69efc8bf0bdf7327339c7b59112dadbb68abb1e (old code)
- 6c04eda2ec5dea6d3b2eae22cafe40caf7cb806b (old code)
- 6c04eda2ec5dea6d3b2eae22cafe40caf7cb806b (already applied)
- 8bdf38b791f1cfe14a4b536c46f4f3e954fa04eb (already applied)
- 40d40cff33ad1c7eaf65d80920b13b76e07b5c8e (already applied)
- f95a2cc0ee8dac7bc4232406afdd82317c1eaf73 (already applied)

ACoquereau committed:
Fix printing of status when in interpretation mode and add clean fmt in the print_fmt function

ACoquereau committed:
Fix new options in output section in the manpage

ACoquereau committed:
Add timeout-as-unknown option

iguerNL committed:
working examples for linear optimization feature

iguerNL committed:
New option `--objectives-in-model` to show optimized expressions in 'model'

If not set, a dedicated section '(objectives ...)' is added for that

iguerNL committed:
Theory: harden 'compute_concrete_model' function

iguerNL committed:
CS: functions return all the env instead of just 'gamma_finite'.

iguerNL committed:
Push the test 'get_case_split_policy() == origin' into Theory.do_case_split

iguerNL committed:
Refactoring to reduce "printing unbounded values" patch in uf.ml.

iguerNL committed:
reintroduce debug traces in satml

iguerNL committed:
Theory.query: simple test to deactivate TCP for both FunSat and CDCL

iguerNL committed:
Fix expr.ml: purify inside Optimize _ as well

iguerNL committed:
begin-optim-alt: Frontend: new syntax for optimisation constraints, with their ordering.

iguerNL committed:
prepare CS to return 'Unbounded' in case of optimization requests.

iguerNL committed:
Infrastructure in theories for CS with optimized values

iguerNL committed:
IntervalCalculus: Produce optimized splits for LIA and LRA

iguerNL committed:
Detect optimization directives (predicates) in Theory and update env

iguerNL committed:
update SMT2 model output with '(objectives ...)' section

iguerNL committed:
satML: add missing case-split strategies

iguerNL committed:
Fun_sat: refactor the way models are handled/printed

- add a data-structure for models,
- save the strucutre in the SAT's env
- print the models in the Frontend module

iguerNL committed:
UF: fix model_extraction: arguments of function may be interpreted

Assertion used as invariant is not accurate

iguerNL committed:
satML: implement models generation.

iguerNL committed:
Print models of arrays (home made syntax)

iguerNL committed:
Models: don't "assert false" when seeing an AC symbols in 'assign_next'

Models gen for AC symbols in currently partial. See the comment in
file ac.ml for more details.

iguerNL committed:
Models: re-introduce removed invariant check

iguerNL committed:
Correctly handle Case-split backtrack when having Optimization constraints

iguerNL committed:
Handling disjunctions in optimization

iguerNL committed:
fix translate ordering in pstm2_to_alt_ergo

iguerNL committed:
Translate minimize/maximize from psmt2-frontend to Alt-Ergo

iguerNL committed:
fix translate from SMT2 to AE

iguerNL committed:
re-implement all-models feature differently

all_models: reset last saved model before exploring other branches

all_models: re-set timelimit, which is inteded to be used per SAT branch/model in this case

print_status: Unknown instead of Timeout if get_interpretation && why3 output

continue models enumeration after a Timeout

avoid infinite loop in case no progression is made in all-models enumeration

do not support check-all-sat/all-models in Fun_SAT

add missing cases to GUI

iguerNL committed:
compute model of terms starting from the smallest one

iguerNL committed:
optim:re-set timelimit when trying to find a better min in other branches

iguerNL committed:
Models: export the map of term |-> get-value(term) computed in UF

Halbaroth committed:
Don't run ocp-indent on mly files. See the issue
https://github.com/OCamlPro/ocp-indent/issues/218

Halbaroth committed:
Fix indentation in the legacy parser

Halbaroth committed:
Fix indentation in the legacy lexer

Halbaroth committed:
Print always smtlib comments in the models

bclement-ocp commented:
This is a huge PR and it will take some time to properly review. Anything after "Add timeout-as-unknown option" (that seems to be the latest functionally-significant commit, at a glance) that is not necessary for the PR to compile and/or tests to pass should be split off into a separate PR to merge after the "main" PR if possible.

> I removed in https://github.com/OCamlPro/alt-ergo/commit/b745947fcb3160ecad029cc5f32cecaf9fc29d11 the why3 output format for models. AE will print only smtlib format for models.
> I am not sure what was the purpose of this format.

This smells of Chesterton's fence. Did you check with the Why3 team and/or with MERCE (who I believe funded this work) that this is not a feature they need before removing it?

Halbaroth committed:
Negative tests

Some tests about models failed in the OptimAE PR.
This commit allows to tag tests in `tests/` with `fail` tag
which means the test is supposed to fail.

Halbaroth committed:
Fix the js worker

Halbaroth committed:
Remove gui artefact

Halbaroth committed:
Rebase artefacts

Halbaroth commented:
I managed to split this PR into two parts. Please don't review it before I push the split ;)

Halbaroth commented:
I will try to extract the CheckSatAll feature from this PR also.
